### ISSUE TITLE
Monitoring events

### ISSUE THREAD

**Issue by Astr0boy** [waarp/WaarpR66#176]
Hi Frédéric,

Is there any way to get via a java interface or any other way all the events notification sent to the Web Administration ?
As I can see that in Transfers, you can have the current step of the transfer even if it has not started.
I would like to catch some errors occurring before the transfer starts (host unknown, auth incorrect...) and looks like it's not possible using error task (I even tried option -Dopenr66.executebeforetransferred=0 but without success).
Any clue or advise ?

**Comment by fredericBregier** [comment]
Hi,
I try first to recall the protocol logic, since it can be useful. (see http://waarp.github.io/Waarp/WaarpR66Protocol.html )

1) Network connection if none already exists between the 2 partners
2) Multiplexing startup (one new query through the same shared socket) : STARTUP packet
=> session is setup
3) Authentication : AUTHENT packet
4) Query request : REQUEST packet 
=> that's where the DbTaskRunner is setup (if not already existing of course for query like restart or status requests)

So Web status are coming from (mainly) from DbTaskRunner. Some are coming from DbHost or related to from step 3 (host nb of connection, remote information as version of partner), and minimal one comes from step 2 (network status).

-Dopenr66.executebeforetransferred=0 is a backward compatibility issue (very old version) since it appears that some errors do not called ERROR tasks when the error comes too early, so the reason we track now this and call whatever the status is the ERROR tasks (or should, might be a bug if you do not have any call). But this call can only be done once step 4 is made since before DbTaskRunner does not exist.
So setting it to 0 will reduce the number of ERROR task called for few rare cases, but not in case of bad authentication. (see details here http://waarp.github.io/Waarp/WaarpR66Options.html section "Error task on Init step transfer").

Various ways exist to know the status outside Waarp R66 Web admin IHM:
1) Getting information post event, from a "pooling" perspective
- using SNMP but in general limited to overall status and integration in a monitoring tool
- HTTP REST support see http://waarp.github.io/Waarp/WaarpR66RestAPI.html but need to use some extra work since need to use the specific REST authentication control (note, this is the way central Waarp Manager is getting information directly, so it is really meaned for that).
- using Thrift interface (similar to REST API)
- using Business command (similar to JavaTask but as specific query of your own which could interact with all R66 servers, based on your own jar). For instance, this is how SpooledDirectory is managed.

2) Getting informations based on events, not pooling yourself
- Using ERROR tasks: in general using a Java task (so extra jar of your own) and you could inspire yourself on how Reschedule is made since it uses the error code to manage the action to do.
- Using a R66BusinessFactory (limited to the event trapped)

I believe this is all (I may forget some other ways of course).
Frederic

**Comment by fredericBregier** [comment]
Of course, we could investigate if some events are not traced for one particular way. For instance, currently R66BusinessInterface does not trace the "startup" (step2) / "authent" (step3) events, but starts at step4 (request).

Do you have some specifications in mind ?

**Comment by fredericBregier** [comment]
You can have a look at a proposal, that I imagine from your request.
Of course, any better idea or complement are welcome !

**Comment by Astr0boy** [comment]
Hi Frédéric,
Thanks for your full detailed answers :)
The protocol logic is really useful, I think I'll print it to have it nearby at any time :)
As per the current limitation of R66BusinessInterface I used some EXECJAVA that I added in the rule. The only thing is that as my java class is used for every transfer, I would rather have a business Interface to manage it and don't have to include it if I need to create other rules.
The ERROR task is working fine, I was just missing a class but it didn't show anywhere so it took me a bit to find out what was wrong... Happiness of development ;)
FYI, we are using the REST API for triggering a transfer between 2 servers (async mode).
Regarding your proposal, I think it's really a plus to add a R66BusinessInterface via the XML config file and have the same events (or even more) than the ERROR task.
Thanks for the good work and your help!
